src/document/json/rht_pq_map.rs

```diff
@@ -0,0 +1,272 @@
+}
+
+impl PartialEq for RHTPQMapNode {
+    fn eq(&self, other: &Self) -> bool {
+        self.key == other.key
+    }
+}
+
+impl Eq for RHTPQMapNode {}
+
+impl Clone for RHTPQMapNode {
+    fn clone(&self) -> RHTPQMapNode {
+        RHTPQMapNode {
+            key: self.key.clone(),
+            element: self.element.clone(),
+        }
+    }
+}
+
+impl RHTPQMapNode {
+    pub fn new(key: String, element: BoxedElement) -> RHTPQMapNode {
+        RHTPQMapNode { key, element }
+    }
+
+    pub fn remove(&self, ticket: Ticket) -> bool {
+        return self.element.remove(ticket);
+    }
+
+    pub fn is_removed(&self) -> bool {
+        match self.element.removed_at() {
+            Some(_) => true,
+            _ => false,
+        }
+    }
+
+    pub fn key(&self) -> String {
+        self.key.clone()
+    }
+}
+
+pub struct RHTPriorityQueueMap {
+    node_queue_map_by_key: HashMap<String, BinaryHeap<Rc<RefCell<RHTPQMapNode>>>>,
+    node_map_by_created_at: HashMap<Ticket, Rc<RefCell<RHTPQMapNode>>>,
+}
+
+impl RHTPriorityQueueMap {
+    pub fn new() -> RHTPriorityQueueMap {
+        RHTPriorityQueueMap {
+            node_queue_map_by_key: HashMap::new(),
+            node_map_by_created_at: HashMap::new(),
+        }
+    }
+
+    pub fn get(&self, key: &str) -> Option<BoxedElement> {
+        let node = match self.node_queue_map_by_key.get(key) {
+            Some(queue) => match queue.peek() {
+                Some(node) => Some(node),
+                _ => None,
+            },
+            _ => None,
+        };
+
+        if node.is_none() {
+            return None;
+        }
+
+        let node = node.unwrap().borrow();
+        match node.is_removed() {
+            false => Some(node.element.clone()),
+            true => None,
+        }
+    }
+
+    pub fn has(&self, key: &str) -> bool {
+        match self.node_queue_map_by_key.get(key) {
+            Some(queue) => match queue.peek() {
+                Some(node) => !node.borrow().is_removed(),
+                _ => false,
+            },
+            _ => false,
+        }
+    }
+
+    pub fn set(&mut self, key: String, value: BoxedElement) -> Option<BoxedElement> {
+        let node = match self.node_queue_map_by_key.get(&key) {
+            Some(queue) => match queue.peek() {
+                Some(node) => Some(node),
+                _ => None,
+            },
+            _ => None,
+        };
+
+        let removed = match node {
+            Some(node) => {
+                let node = node.borrow();
+                if node.is_removed() {
+                    return None;
+                }
+
+                if node.remove(value.created_at()) {
+                    return Some(node.element.clone());
+                } else {
+                    return None;
+                }
+            }
+            _ => None,
+        };
+
+        self.set_internal(key, value);
+
+        removed
+    }
+
+    pub fn set_internal(&mut self, key: String, value: BoxedElement) {
```

Comment:
```suggestion
    fn set_internal(&mut self, key: String, value: BoxedElement) {
```
This method seems like the private method.